compiler/src/iree/compiler/Codegen/Dialect/PCF/Transforms/FuseConsumers.cpp

```diff
@@ -0,0 +1,916 @@
+// Consumer fusion impls
+//===---------------------------------------------------------------------===//
+
+template <typename OpTy>
+static LogicalResult
+lookupProducerSlices(OpResult result,
+                     SmallVectorImpl<PCF::WriteSliceOp> &slices) {
+  OpTy owner = cast<OpTy>(result.getOwner());
+  Value tiedArg = owner.getRegionRefArgs()[result.getResultNumber()];
+  auto srefType = cast<PCF::ShapedRefType>(tiedArg.getType());
+  if (!srefType.isReturnOnlySync()) {
+    return failure();
+  }
+  for (auto user : tiedArg.getUsers()) {
+    // We can ignore memory reads.
+    if (isa<PCF::ReadSliceOp>(user)) {
+      continue;
+    }
+    auto sliceOp = dyn_cast<PCF::WriteSliceOp>(user);
+    // TODO: Support vector operands.
+    if (!sliceOp || !isa<RankedTensorType>(sliceOp.getSourceType()) ||
+        !sliceOp.hasUnitStride()) {
+      return failure();
+    }
+    slices.push_back(sliceOp);
+  }
+  return success();
+}
+
+// Two cases: one operand with multiple producer slices or multiple operands
+// with a single producer slice per operand.
+// Currently multiple operands <-> multiple producers is unsupported.
+template <typename OpTy>
+static LogicalResult
+matchTilableConsumerImpl(RewriterBase &rewriter, OpTy producerOp,
+                         TilingInterface target, ConsumerFusionParams &params) {
+  // To create a loop result we need either an initializer or a shape. This
+  // can come from either ReifyRankedShapedTypeOpInterface or DPS. If the
+  // operand being fused along is itself a destination we get the shape via
+  // passthrough with the producer's init/shape.
+  if (!isa<ReifyRankedShapedTypeOpInterface, DestinationStyleOpInterface>(
+          *target)) {
+    return rewriter.notifyMatchFailure(
+        target, "unsupported non-reify result shapes or dps op");
+  }
+
+  SmallVector<unsigned> &targetOperands = params.operands;
+  SetVector<unsigned> &targetResults = params.results;
+  SmallVector<PCF::WriteSliceOp> &slices = params.slices;
+  assert(targetOperands.empty() && "unexpected non-empty operand list");
+  assert(targetResults.empty() && "unexpected non-empty result set");
+  assert(slices.empty() && "unexpected non-empty slice list");
+  mlir::DominanceInfo dominanceInfo(producerOp->getParentOp());
+  // First collect the set of operands/results fused along. Additionally verify
+  // dominance for other operands.
+  for (OpOperand &operand : target->getOpOperands()) {
+    auto opResult = dyn_cast<OpResult>(operand.get());
+    if (opResult && opResult.getOwner() == producerOp) {
+      targetOperands.push_back(operand.getOperandNumber());
+      targetResults.insert(opResult.getResultNumber());
+    } else {
+      if (!dominanceInfo.dominates(operand.get(), producerOp)) {
+        return rewriter.notifyMatchFailure(
+            target, "unable to fuse due to operand dominance");
+      }
+    }
+  }
+
+  if (targetOperands.empty()) {
```

Comment:
I dont know if I should return a failure here. I know this makes sense within a pattern rewriter, but as a stand-alone method, this should just return success.

Reply:
Failure is what I would expect to return here. This matcher says "Is there at least one target consumer to fuse into this loop" which there are not if there are no valid fusable consumers.